### PR TITLE
Raster masking

### PR DESCRIPTION
**Pull request by laurence-hudson-tessella**

### CODE REVIEW

File: src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterTransformation.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * An interface to define a raster transformation operation. This will likely involve iterating through the pixels in
+ * the WritableRaster and setting new values where required for the operation being implemented.
+ *
+ * Copyright (c) 2014 University of Oxford
+ */
+public interface RasterTransformation {
+    /**
+     * Perform a raster transformation operation.
+     * @param raster The raster which should be updated.
+     * @param referenceRasters One or more reference raster which may be compared against when updating the main raster.
```
Comment (edwiles) [inline src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterTransformation.java]: rasters

File: src/ModelOutputHandler/src/uk/ac/ox/zoo/seeg/abraid/mp/modeloutputhandler/web/MainController.java [hunk 2]
```diff
@@ -65,13 +67,26 @@ public MainController(
...
+            // Delete old pre-mask raster outputs (outside of the "mainHandler.handleOutputs" transaction)
+            boolean deleteResult = mainHandler.handleOldRasterDeletion(modelRun.getDiseaseGroupId());
+            if (!deleteResult) {
+                sendCleanUpFailureEmail(modelRun);
+            }
+
+            // Announce failed model runs
             if (modelRun.getStatus() == ModelRunStatus.FAILED) {
                 sendModelFailureEmail(modelRun);
+
```
Comment (edwiles) [inline src/ModelOutputHandler/src/uk/ac/ox/zoo/seeg/abraid/mp/modeloutputhandler/web/MainController.java]: Extra blank line

File: src/ModelOutputHandler/src/uk/ac/ox/zoo/seeg/abraid/mp/modeloutputhandler/web/MainHandler.java [hunk 3]
```diff
@@ -100,13 +108,45 @@ public ModelRun handleOutputs(File modelRunZipFile) throws ZipException, IOExcep
...
+    /**
+     * Handles the deletion of old model outputs. These are limited to pre-masking raster files that are no longer the
+     * newest model run, for the given disease group.
+     * @param diseaseGroupId The given disease group's id.
+     * @return A boolean which is true if the deletions succeeded,
+     *         or is false if one or more files could not be deleted (e.g. are currently in use elsewhere).
+     */
+    public boolean handleOldRasterDeletion(int diseaseGroupId) {
+        ModelRun keep = modelRunService.getMostRecentlyFinishedModelRunWhichCompleted(diseaseGroupId);
+        Collection<ModelRun> delete = modelRunService.getModelRunsForDiseaseGroup(diseaseGroupId);
```
Comment (edwiles) [inline src/ModelOutputHandler/src/uk/ac/ox/zoo/seeg/abraid/mp/modeloutputhandler/web/MainHandler.java]: runsToDelete?

File: src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterUtils.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    private static final Hints RASTER_READ_HINTS =
+            new Hints(Hints.DEFAULT_COORDINATE_REFERENCE_SYSTEM, GeometryUtils.WGS_84_CRS);
+
+    /**
+     * The standard raster NODATA value (-9999)
+     */
+    public static final int NO_DATA_VALUE = -9999;
+
+    /**
+     * A alternative raster NODATA value, to be used instead of RasterUtils.NO_DATA_VALUE for pixel where the value is
```
Comment (edwiles) [inline src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterUtils.java]: pixels

**Comment by helena-patching-tessella** [comment]
LGTM

File: config/geoserver/abraid/styles/abraid_raster.sld [hunk 5]
```diff
@@ -15,6 +15,7 @@
                             <sld:ColorMapEntry color="#ffffcb" opacity="1" quantity="0.5"/>
                             <sld:ColorMapEntry color="#cf93ba" opacity="1" quantity="0.75"/>
                             <sld:ColorMapEntry color="#a44883" opacity="1" quantity="1.0"/>
+                            <sld:ColorMapEntry color="#eaeaea" opacity="1" quantity="+9999"/>
```
Comment (edwiles) [inline config/geoserver/abraid/styles/abraid_raster.sld]: Obviously the + is not strictly necessary but I presume you're using it to clearly distinguish from -9999?

File: src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterTransformation.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.awt.image.Raster;
+import java.awt.image.WritableRaster;
+import java.io.IOException;
+
+/**
+ * An interface to define a raster transformation operation. This will likely involve iterating through the pixels in
+ * the WritableRaster and setting new values where required for the operation being implemented.
+ *
+ * Copyright (c) 2014 University of Oxford
```
Comment (edwiles) [inline src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterTransformation.java]: New files created in 2015 should be copyright 2015

File: src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterTransformation.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ */
+public interface RasterTransformation {
+    /**
+     * Perform a raster transformation operation.
+     * @param raster The raster which should be updated.
+     * @param referenceRasters One or more rasters which may be compared against when updating the main raster.
+     * @throws IOException thrown if the transformation can not be completed.
```
Comment (edwiles) [inline src/Common/src/uk/ac/ox/zoo/seeg/abraid/mp/common/util/RasterTransformation.java]: cannot

**Comment by edwiles** [comment]
LGTM with tiny comments

**Pull request merged**